runtime/vm/resolvefield.cpp

```diff
@@ -578,6 +583,124 @@ addHiddenInstanceField(J9JavaVM *vm, const char *className, const char *fieldNam
+	} else if (clazz->hotFieldsInfo->hotFieldListLength < MAX_HOT_FIELD_LIST_LENGTH) {
+		addOrUpdateHotField(clazz, fieldOffset, reducedCpuUtil, reducedFrequency);
+	}
+}
+
+/**
+ * Create and initialize a class' hotFieldsInfo pool element.
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param javaVM[in] pointer to the J9JavaVM
+ * @param clazz pointer to the class where a hot field should be added
+ * @param fieldOffset value of the field offset that should be added as a hot field for the given class
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+VMINLINE void
+createClassHotFieldsInfo(J9JavaVM *javaVM, J9Class* clazz, U_8 fieldOffset, int32_t reducedCpuUtil, U_32 reducedFrequency)
+{
+	omrthread_monitor_enter(javaVM->hotFieldClassInfoPoolMutex);
+	if (NULL == clazz->hotFieldsInfo) {
+		/* initialize new hotFieldsInfo pool element as it does not exist already */
+		clazz->hotFieldsInfo = (J9ClassHotFieldsInfo *)pool_newElement(javaVM->hotFieldClassInfoPool);
+		if (NULL == clazz->hotFieldsInfo) {
+			goto exit;
+		}
+		clazz->hotFieldsInfo->hotFieldListLength = 0;
+		clazz->hotFieldsInfo->consecutiveHotFieldSelections = 0;
+		clazz->hotFieldsInfo->hotFieldOffset1 = U_8_MAX;
+		clazz->hotFieldsInfo->hotFieldOffset2 = U_8_MAX;
+		
+		/* create and initialize the new hot field */
+		omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);
+		J9HotField* newHotField = (J9HotField *)pool_newElement(clazz->classLoader->hotFieldPool);
+		if(NULL == newHotField) {
+			omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+			goto exit;
+		}
+		newHotField->next = NULL;
+		newHotField->hotFieldOffset = fieldOffset;
+		newHotField->hotness = (reducedFrequency * reducedCpuUtil);
+		newHotField->cpuUtil = reducedCpuUtil;
+		omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+		
+		/* update hotFieldsInfo for the class now that a new hot field has been added */
+		clazz->hotFieldsInfo->hotFieldListHead = newHotField;
+		clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+		clazz->hotFieldsInfo->hotFieldListLength++;
+		clazz->hotFieldsInfo->classLoader = clazz->classLoader;
+	}
+	
+exit:
+	omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+}
+
+/**
+ * Add or update an existing hot field for a given class.
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param clazz pointer to the class where a hot field should be added/updated
+ * @param fieldOffset value of the field offset that should be added/updated as a hot field for the given class
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+VMINLINE void
```

Comment:
might be worth to return a success/err return code here as well. This way you can propogate the err to the JIT thread if you decide to do something with that info.

All allocation failures should set the err.

Reply:
Ok. Shall I will do the basic -1 on failure and 1 on success as I see done in several places in the JIT @andrewcraik?

Reply:
You should use whatever signalling mechanism is most sensible. For the JIT because the data is all optional you can just skip populating the data right? The VM is probably not long for this world with an OOM looming, but there is no reason to abort the compile or bring down the VM - this is a purely optional activity...

Reply:
@tajila would you prefer I leave it to return void or from a VM perspective would you like that I return a success/error return code?

Reply:
Based on Andrew's answer you can leave it as void since reporting is optional. You just need to add a null check in `addOrUpdateHotField` for `clazz->hotFieldsInfo`.